astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
-                result.uncertainty = method(operand_scaled, result.data)
-            except IncompatibleUncertaintiesException:
-                raise IncompatibleUncertaintiesException(
-                    "Cannot propagate uncertainties of type {0:s} with "
-                    "uncertainties of type {1:s} for {2:s}".format(
-                        self.uncertainty.__class__.__name__,
-                        operand.uncertainty.__class__.__name__,
-                        name))
+            # As with self.uncertainty is None but the other way around.
+            operand.uncertainty = self.uncertainty.__class__(None)
+            result_uncert = self.uncertainty.propagate(operation, operand,
+                                                       result, correlation)
+            operand.uncertainty = None
+            return result_uncert
+
+        else:
+            # Both have uncertainties so just propagate.
+            return self.uncertainty.propagate(operation, operand, result,
+                                              correlation)
+
+    def _arithmetic_mask(self, operation, operand, handle_mask, **kwds):
+        """
+        Calculate the resulting mask
+
+        This is implemented as the piecewise ``or`` operation if both have a
+        mask.
+
+        Parameters
+        ----------
+        operation : callable
+            see `NDArithmeticMixin._arithmetic` parameter description.
+
+        operand : `NDData`-like instance
+            The second operand wrapped in an instance of the same class as
+            self.
+
+        handle_mask : callable
+            see :meth:`NDArithmeticMixin.add`
+
+        kwds :
+            Additional parameters given to ``handle_mask``.
+
+        Returns
+        -------
+        result_mask : `bool` or `~numpy.ndarray` or None
+            If only one mask was present this mask is returned.
+            If neither had a mask ``None`` is returned. Otherwise
+            ``handle_mask`` must create the returned mask.
+        """
 
+        # If only one mask is present we need not bother about any type checks
         if self.mask is None and operand.mask is None:
-            result.mask = None
+            return None
         elif self.mask is None:
-            result.mask = operand.mask.copy()
+            # Make a copy so there is no reference in the result.
+            return deepcopy(operand.mask)
         elif operand.mask is None:
-            result.mask = self.mask.copy()
-        else:  # combine masks as for Numpy masked arrays
-            result.mask = self.mask | operand.mask  # copy implied by operator
+            return deepcopy(self.mask)
+        else:
+            # Now lets calculate the resulting mask (operation enforces copy)
+            return handle_mask(self.mask, operand.mask, **kwds)
 
-        return result
+    def _arithmetic_wcs(self, operation, operand, compare_wcs, **kwds):
```

Comment:
Yes, I included it simply to have the same method signature for wcs, mask and meta. I included that it is ignored.